File: src/turingos/envelope.py
"""turingos.envelope — the 7-field Append Envelope (B-2, ADR-0002) [Art. 0.2 / 0.3].

Frozen Stage-0 interface (contracts/INTERFACES.md envelope.py section,
contracts/append_envelope.md). One micro Git commit per sovereignty-boundary change carries
exactly one event whose envelope is this 7-field shape. Five fields are load-bearing in 1.0,
two are reserved forward-compat seams (`writer_id` fixed single-writer; `authority_epoch`
deferred — no fencing in 1.0).

This module is a pure data carrier + payload projection. It does NOT touch git, the codec, or
the Tape; the local guard checks (FF parent, schema-known, head_effect-registry-derived,
payload_hash match, single-writer identity, accepted_head ancestor) live on the Tape's append
path. The one piece of policy enforced here is that head_effect is REGISTRY-DERIVED, never
writer-trusted: callers obtain it via derive_head_effect(), which delegates to registry.head_effect.
"""
from __future__ import annotations

from dataclasses import dataclass

from . import registry


@dataclass(frozen=True)
class AppendEnvelope:
    """The frozen 7-field append envelope.

    Field order is part of the frozen contract (positional construction is supported):

      prev_tape_tip        load-bearing — FF parent: parent == prev_tape_tip (reject non-FF)
      event_schema_id      load-bearing — event type / payload schema id (closed-world)
      payload_hash         load-bearing — = content_digest = sha256(JCS(payload))
      head_effect          load-bearing — registry-derived (ADVANCE|PRESERVE); never writer-trusted
      accepted_head_before load-bearing — accepted_head the writer observed (ancestor-of tape_tip)
      writer_id            RESERVED-fixed — single active sovereign writer (audit/handoff)
      authority_epoch=0    RESERVED-deferred — not enforced in 1.0 (no lease/epoch fencing)
    """

    prev_tape_tip: str
    event_schema_id: str
    payload_hash: str
    head_effect: str
    accepted_head_before: str
    writer_id: str
    authority_epoch: int = 0

    def to_payload(self) -> dict:
        """Project the envelope to a plain dict with exactly the 7 ASCII keys.

        A fresh dict each call (mutating the result never affects the frozen envelope), so it is
        safe to feed straight to the JCS codec / commit builder.
        """
        return {
            "prev_tape_tip": self.prev_tape_tip,
            "event_schema_id": self.event_schema_id,
            "payload_hash": self.payload_hash,
            "head_effect": self.head_effect,
            "accepted_head_before": self.accepted_head_before,
            "writer_id": self.writer_id,
            "authority_epoch": self.authority_epoch,
        }


def derive_head_effect(event_type: str) -> str:
    """Return 'ADVANCE' | 'PRESERVE' for an event type, REGISTRY-DERIVED.

    Thin delegation to registry.head_effect — the single source of truth for ADVANCE/PRESERVE.
    Carrying head_effect via this function (rather than trusting a writer-supplied value) is what
    makes the Tape guard able to REJECT any envelope whose head_effect disagrees with the registry.
    An unknown event_type propagates registry's closed-world errors.RejectedAppend.
    """
    return registry.head_effect(event_type)

File: crates/turing-contracts/src/envelope.rs
//! `MicroEventEnvelope.v1` — the committed Git body — and `HeadSet.v1`.
//!
//! These are the closed Rust types for `pack/05_schemas/micro_event_envelope.v1.schema.json`
//! and `pack/05_schemas/head_set.v1.schema.json`. They serialize to the **exact**
//! committed body via the `turingos.jcs.v1` codec: the canonical bytes of an
//! [`MicroEventEnvelope`] are what a Git commit stores, and `content_digest ==
//! payload_hash == sha256:` + `sha256(JCS(payload))`.
//!
//! Two frozen invariants are encoded in the type itself:
//! - **`event_id` is forbidden in the committed body** — a Git commit cannot embed its
//!   own OID; the external read identity is paired with the body through
//!   `micro_event_record.v1` and the append receipt. There is therefore no `event_id`
//!   field anywhere on [`MicroEventEnvelope`].
//! - **`head_set_after` is forbidden** — the post-state is self-reference-free and is
//!   *derived* (see [`crate::registry`] + the kernel reducer), never carried.
//!
//! `head_effect` is carried for audit only; an admitting kernel accepts it solely when
//! it equals the registry row (`pack/04_registries/event_registry_v5_3_1.json`). This
//! module does not trust it.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::jcs::{self, JcsError};

/// The frozen `head_effect` discriminator: whether an event *may* advance its registry
/// head. The actual movement also requires `predicate_product == PASS` and is decided
/// by the kernel reducer; this is the registry-declared intent.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum HeadEffect {
    /// The event's registry head advances iff `predicate_product == PASS`.
    #[serde(rename = "ADVANCE")]
    Advance,
    /// The event never advances any sovereign head; only `tape_tip` moves.
    #[serde(rename = "PRESERVE")]
    Preserve,
}

/// The frozen predicate product: the deterministic boolean (or "not applicable") result
/// of the Candidate Predicate over frozen Tape bytes.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum PredicateProduct {
    /// The transition is verified; a registry head may advance.
    #[serde(rename = "PASS")]
    Pass,
    /// The transition failed; only a failure node lands on `tape_tip`.
    #[serde(rename = "FAIL")]
    Fail,
    /// The event is predicate-free (`predicate_required == false`); no head advances.
    #[serde(rename = "NOT_RUN")]
    NotRun,
}

/// `MicroEventEnvelope.v1` — the 16-field committed Git body.
///
/// Field order in this struct mirrors the schema's `required` array for readability; the
/// committed bytes are produced by the `turingos.jcs.v1` codec which sorts keys bytewise,
/// so the struct field order is not load-bearing. `serde` field names are the exact
/// schema keys and there is **no** `event_id` / `head_set_after` field
/// (`additionalProperties:false` parity is enforced at parse by [`Self::from_jcs_value`]).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MicroEventEnvelope {
    /// Always the const `"micro_event_envelope.v1"`.
    pub schema_id: String,
    /// The canonical registry event name (e.g. `"SystemConstitutionAccepted"`).
    pub event_type: String,
    /// The opaque writer identity that minted this candidate.
    pub writer_id: String,
    /// The pre-state fencing epoch (`>= 0`).
    pub authority_epoch: u64,
    /// The monotone per-Tape sequence number (`>= 0`).
    pub sequence: u64,
    /// The observed `tape_tip` this event is a child of (`mu:`…), or `null` at genesis.
    pub prev_tape_tip: Option<String>,
    /// The `authorization_head` observed pre-append (`mu:`…), or `null`.
    pub authorization_head_before: Option<String>,
    /// The `accepted_head` observed pre-append (`mu:`…), or `null` at genesis.
    pub accepted_head_before: Option<String>,
    /// Carried-for-audit head effect; accepted only if equal to the registry row.
    pub head_effect: HeadEffect,
    /// The payload's schema id (e.g. `"system_constitution_accepted.v1"`).
    pub event_schema_id: String,
    /// The predicate product for this final event.
    pub predicate_product: PredicateProduct,
    /// `sha256:` + 64hex over the sorted predicate reasons (`sha256(JCS([]))` when none).
    pub reason_digest: String,
    /// Whether this event is a verified transition (`true` only on PASS finalization).
    pub verified: bool,
    /// `sha256:` + `sha256(JCS(payload))`; must equal `payload_hash`.
    pub content_digest: String,
    /// Append-API name for `content_digest`; byte-equal to it.
    pub payload_hash: String,
    /// The event payload object (schema is `event_schema_id`).
    pub payload: Value,
}

/// The const `schema_id` for the committed envelope body.
pub const ENVELOPE_SCHEMA_ID: &str = "micro_event_envelope.v1";

const ENVELOPE_REQUIRED_FIELDS: &[&str] = &[
    "schema_id",
    "event_type",
    "writer_id",
    "authority_epoch",
    "sequence",
    "prev_tape_tip",
    "authorization_head_before",
    "accepted_head_before",
    "head_effect",
    "event_schema_id",
    "predicate_product",
    "reason_digest",
    "verified",
    "content_digest",
    "payload_hash",
    "payload",
];

impl MicroEventEnvelope {
    /// The canonical `turingos.jcs.v1` bytes of this envelope — exactly what a Git
    /// commit stores as the event body. Sorted keys, no whitespace, integers only.
    pub fn to_jcs_bytes(&self) -> Result<Vec<u8>, JcsError> {
        let value = serde_json::to_value(self)
            .expect("MicroEventEnvelope serializes to a JSON object infallibly");
        jcs::canonicalize(&value)
    }

    /// Parse a committed body from already-loaded JSON, enforcing the schema's closed
    /// shape: the const `schema_id`, exactly the 16 fields, and the absence of
    /// `event_id` / `head_set_after`. `serde(deny_unknown_fields)` rejects any extra key
    /// (including those two), so a torn or tampered body fails here rather than silently
    /// round-tripping.
    pub fn from_jcs_value(value: &Value) -> Result<Self, JcsError> {
        let obj = value
            .as_object()
            .ok_or_else(|| JcsError::Malformed("envelope body is not a JSON object".into()))?;
        if obj.contains_key("event_id") {
            return Err(JcsError::ForbiddenPayloadField("event_id".into()));
        }
        if obj.contains_key("head_set_after") {
            return Err(JcsError::ForbiddenPayloadField("head_set_after".into()));
        }
        for field in ENVELOPE_REQUIRED_FIELDS {
            if !obj.contains_key(*field) {
                return Err(JcsError::Malformed(format!(
                    "envelope shape: missing required field {field:?}"
                )));
            }
        }
        if obj.len() != ENVELOPE_REQUIRED_FIELDS.len() {
            return Err(JcsError::Malformed(format!(
                "envelope shape: expected exactly {} fields, got {}",
                ENVELOPE_REQUIRED_FIELDS.len(),
                obj.len()
            )));
        }
        let env: MicroEventEnvelope = serde_json::from_value(value.clone())
            .map_err(|e| JcsError::Malformed(format!("envelope shape: {e}")))?;
        if env.schema_id != ENVELOPE_SCHEMA_ID {
            return Err(JcsError::Malformed(format!(
                "schema_id must be {ENVELOPE_SCHEMA_ID:?}, got {:?}",
                env.schema_id
            )));
        }
        Ok(env)
    }
}

/// `HeadSet.v1` — the coherent three-ref sovereign state.
///
/// `tape_tip` and `accepted_head` are always non-null once genesis exists;
/// `authorization_head` is nullable (no AUTHORIZATION event has been accepted yet). This
/// type carries **no** epoch/sequence field — those are reconciled from the envelope
/// pre-state (schema parity: `additionalProperties:false`).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct HeadSet {
    /// The tip of the Tape (`mu:`…); advances on every valid append.
    pub tape_tip: String,
    /// The latest accepted AUTHORIZATION event (`mu:`…), or `null`.
    pub authorization_head: Option<String>,
    /// The latest accepted SOVEREIGN_ACCEPT event (`mu:`…); non-null post-genesis.
    pub accepted_head: String,
}

/// The const `schema_id` value for `head_set.v1` consumers (the schema itself has no
/// `schema_id` field; this is the registry identifier used in prose/receipts).
pub const HEAD_SET_SCHEMA_ID: &str = "head_set.v1";
